### PR TITLE
Improve datatypes

### PR DESCRIPTION
**Pull request by bastelfreak**

### CODE REVIEW

File: manifests/init.pp [hunk 1]
```diff
@@ -1,64 +1,64 @@
...
-  Hash[String, Hash] $hosts                               = {},
-  Hash[String, Hash] $ignoredsubnets                      = {},
-  Hash[String, Hash] $pools                               = {},
-  Hash[String, Hash] $pools6                              = {},
-  Optional[Stdlib::Absolutepath] $dhcpd_binary            = $dhcp::params::dhcpd_binary
+  Optional[Array[String[1]]] $dnsdomain                            = undef,
+  Array[Stdlib::IP::Address::V4] $nameservers                      = [],
+  Array[Stdlib::IP::Address::V6] $nameservers_ipv6                 = [],
+  Array[Variant[Stdlib::Fqdn,Stdlib::IP::Address]] $ntpservers     = [],
+  Array[Stdlib::Fqdn] $dnssearchdomains                            = [],
```
Comment (alexjfisher) [inline manifests/init.pp]: I think this is too restrictive as it stops putting a top level domain in.

File: manifests/init.pp [hunk 2]
```diff
@@ -1,64 +1,64 @@
...
+  String[1] $dhcp_conf_header                                      = 'INTERNAL_TEMPLATE',
+  String[1] $dhcp_conf_ddns                                        = 'INTERNAL_TEMPLATE',
+  String[1] $dhcp_conf_ntp                                         = 'INTERNAL_TEMPLATE',
```
Comment (alexjfisher) [inline manifests/init.pp]: Are empty strings definitely not allowed here?

File: manifests/init.pp [hunk 3]
```diff
@@ -1,64 +1,64 @@
...
+  String[1] $interface                                             = 'NOTSET',
+  Optional[String[1]] $dnsupdatekey                                = undef,
+  String[1] $ddns_update_style                                     = 'interim',
+  Optional[String[1]] $dnskeyname                                  = undef,
+  String[1] $ddns_update_static                                    = 'on',
+  String[1] $ddns_update_optimize                                  = 'on',
+  Enum['allow', 'deny'] $ddns_client_updates                       = 'allow',
+  Optional[Stdlib::IP::Address] $pxeserver                         = undef,
+  Optional[String[1]] $pxefilename                                 = undef,
+  Optional[Integer[1]] $mtu                                        = undef,
```
Comment (alexjfisher) [inline manifests/init.pp]: An MTU of `1` is no more valid than `0`

File: manifests/init.pp [hunk 4]
```diff
@@ -1,64 +1,64 @@
...
+  Optional[String[1]] $ipxe_filename                               = undef,
+  Optional[String[1]] $ipxe_bootstrap                              = undef,
+  String[1] $logfacility                                           = 'daemon',
```
Comment (alexjfisher) [inline manifests/init.pp]: Can an `Enum` be used here?

File: manifests/init.pp [hunk 5]
```diff
@@ -1,64 +1,64 @@
...
+  String[1] $dhcpd_conf_filename                                   = 'dhcpd.conf',
+  String[1] $packagename                                           = $dhcp::params::packagename,
+  Boolean $manage_package                                          = true,
+  Variant[String[1],Array[String[1]]] $servicename                 = $dhcp::params::servicename,
+  Boolean $manage_service                                          = true,
+  Optional[String[1]] $package_provider                            = $dhcp::params::package_provider,
+  Stdlib::Port $ldap_port                                          = 389,
+  String[1] $ldap_server                                           = 'localhost',
+  String[1] $ldap_username                                         = 'cn=root, dc=example, dc=com',
+  String[0] $ldap_password                                         = '',
```
[the remark this answers is not in the document]
Comment (bastelfreak) [reply manifests/init.pp]: the default is an empty string and I didn't want to change it. Personally I wuld perfer undef. Do you want me to change it? That would be a breaking change (next release already is a major one, so we could include it).

File: manifests/init.pp (hunk 1 above)
Comment (bastelfreak) [reply manifests/init.pp]: will update it to `String[1]` than.

File: manifests/init.pp (hunk 3 above)
Comment (bastelfreak) [reply manifests/init.pp]: I'm not sure which is the lowest positive number that makes sense, I just wanted to prohibit negative numbers.

File: manifests/init.pp (hunk 4 above)
Comment (bastelfreak) [reply manifests/init.pp]: that was my idea as well, but I wasn't able to find a list of allowed values. Is that something that's defined in the syslog RFC? Can you create custom facilities as well? :thinking:

File: manifests/init.pp (hunk 5 above)
[the remark this answers is not in the document]
Comment (alexjfisher) [reply manifests/init.pp]: I meant why not leave it as `String`

File: manifests/init.pp (hunk 4 above)
Comment (alexjfisher) [reply manifests/init.pp]: I’ve got https://github.com/voxpupuli/puppet-nrpe/blob/master/types/syslogfacility.pp

File: manifests/init.pp (hunk 2 above)
Comment (bastelfreak) [reply manifests/init.pp]: allowed empty string again

File: manifests/init.pp (hunk 5 above)
[the remark this answers is not in the document]
Comment (bastelfreak) [reply manifests/init.pp]: ah yes. doesn't make a difference

File: manifests/init.pp (hunk 4 above)
Comment (bastelfreak) [reply manifests/init.pp]: updated this PR to use the datatype and raised an issue here and created https://github.com/puppetlabs/puppetlabs-stdlib/pull/1005

**Pull request merged**